lgc/patch/PatchBoundsCheckMemory.cpp

```diff
@@ -0,0 +1,195 @@
+/*
+ ***********************************************************************************************************************
+ *
+ *  Copyright (c) 2020 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+ ***********************************************************************************************************************
+ * @file  PatchBoundsCheckMemory.cpp
+ * @brief LLPC source file: contains implementation of class lgc::PatchBoundsCheckMemory.
+ ***********************************************************************************************************************
+ */
+#include "PatchBoundsCheckMemory.h"
+#include "lgc/state/IntrinsDefs.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/raw_ostream.h"
+
+#define DEBUG_TYPE "lgc-bounds-check-memory"
+
+using namespace llvm;
+using namespace lgc;
+
+namespace lgc {
+
+// =====================================================================================================================
+// Initializes static members.
+char PatchBoundsCheckMemory::ID = 0;
+
+// =====================================================================================================================
+// Pass creator, creates the pass of bounds check memory operations.
+ModulePass *createPatchBoundsCheckMemory() {
+  return new PatchBoundsCheckMemory();
+}
+
+static Value *getTypeZero(Type *ty) {
+  if (ty->isIntegerTy()) {
+    return ConstantInt::get(ty, 0);
+  } else if (ty->isFloatTy() || ty->isHalfTy() || ty->isDoubleTy() || ty->isFP128Ty() || ty->isBFloatTy()) {
+    return ConstantFP::get(ty, 0.0);
+  } else {
```

Comment:
Remove else. Can we assert that `ty` is an aggregate type?

Reply:
There is such an assertion right at the top of `ConstantAggregateZero::get`. I’d rather not copy it here because it checks for three different types.